### PR TITLE
Issue 112/135 caching memoise api get

### LINKED ISSUE #78: Support `versions` argument in get_file/get_dataframe
For datasets that periodically update (e.g. [this](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/42MVDX) is on V6), it would be good for the `get_*` functions to have an option to specify the version number.

Looks like there is an API for that
https://guides.dataverse.org/en/latest/api/dataaccess.html#download-by-dataset-by-version

~However, see #27 for an error in `dataset_versions`.~

### LINKED ISSUE #112: Faster JSON parser
For dataset retrieval, we download and parse JSON metadata multiple times. 
For example, in `get_dataframe_by_name`, `get_fileid.character` would first find the dataset id via https://github.com/IQSS/dataverse-client-r/blob/4775a92360569adb6e693ad6db940f89530eeb8d/R/utils.R#L29
 and the file the list of ids for each file in the dataset at 
 https://github.com/IQSS/dataverse-client-r/blob/4775a92360569adb6e693ad6db940f89530eeb8d/R/get_dataset.R#L101
 
 It turns out the time this takes is non-trivial. Most of the time is taken up by loading the JSON from URL. A small remaining fraction (< 1%) is due to the parsing of the JSON file. We could make a minor improvement in speed by switching to a faster parser, RcppSimdJson (https://github.com/eddelbuettel/rcppsimdjson). This is about 2-10x faster in my tests, per below.  The current    `jsonlite::fromJSON` seems to be optimal for data science pipelines where we deal with data but here we are only interested in bits of metadata.  An even faster switch is to download metadata only once.
 
 Switching packages will require changes in at least 20 places where jsonlite is used.

``` r
library(jsonlite) # currently used
library(RcppSimdJson) # potential replacement

# sample: https://demo.dataverse.org/file.xhtml?persistentId=doi:10.70122/FK2/PPIAXE/MHDB0O
js_url <- "https://demo.dataverse.org/api/datasets/export?exporter=dataverse_json&persistentId=doi%3A10.70122/FK2/PPIAXE"

# download once
tmp <- tempfile()
download.file(js_url, tmp)

microbenchmark::microbenchmark(
  statusquo = jsonlite::fromJSON(js_url), # what is currently being called
  dl = curl::curl_download(js_url, tempfile()), # separating download from parsing
  jsonlite = jsonlite::fromJSON(tmp),  # parsing, without download
  RcppJson = RcppSimdJson::fload(tmp), # replace with Rcpp
  RcppJson_file = RcppSimdJson::fload(tmp, query = "/datasetVersion/files"), # only files data
  RcppJson_id = RcppSimdJson::fload(tmp, query = "/id"),  # stop at dataset /id
  times = 30
)
#> Unit: microseconds
#>           expr        min         lq        mean      median         uq        max neval
#>      statusquo 365097.709 371235.626 374774.8021 373752.4175 378357.084 387006.459    30
#>             dl 361154.168 364100.750 369091.1201 369528.3965 371835.459 378629.667    30
#>       jsonlite   1487.834   2743.500   3248.0424   2994.1465   3270.959   8380.876    30
#>       RcppJson    186.876    262.001    438.1298    345.3130    468.042   2335.417    30
#>  RcppJson_file    136.292    224.001    334.5173    301.6465    409.376    688.001    30
#>    RcppJson_id    138.459    177.876    287.7714    263.3965    362.792    586.750    30
```

<sup>Created on 2022-01-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### LINKED ISSUE #135: Adding caching to `get_file_by_id`
It would be useful if the package were able to cache results of calls to the DataVerse API to disk.  If the same dataset is requested twice, the result can then be served from the disk instead of re-downloading which would save a lot of time. 

Here's a sketch of how I think the behavior could work: 

### Suggested new behavior for `get_file_by_id`: 

1. Check we are asking to download a specific version of a file. If we are, proceed to the next step. If we are asking for the latest version, the result cannot be cached as the latest version might change between function invocations. 
2. Check if caching is disabled through some sort of Environment variable. If it is, don't cache results either.  
3. Download the file and cache the results. 

More complex behavior could be added on in the future such as caching the latest version by checking if the file metadata has changed since the last download, and only re-downloading if there is a new version. 

I am prototyping the behavior in the cache branch, and would love feedback on the behavior. Right now, I am caching all calls to `get_file_id` which have a specific version of the file specified. I will shortly add step 2 which checks if the user wants to turn off caching if we think this is a good way to go. 

Best, 
Ben

### PR DESCRIPTION
**Pull request by mtmorgan**
This builds on a proposal in a #112 comment (by caching API calls, rather than faster JSON parsing) and closes #135 (`get_file_by_id()` is a particular example of an API call). It also closes #78 with additional documentation. 

The first three commits are made redundant with pull request #136, and would be removed before final merge.

Some comments

- caching is enabled by default. All functions making GET calls can add `use_cache = FALSE` to disable the cache for that call, or `Sys.setenv(DATAVERSE_USE_CACHE = FALSE)` to disable use of the cache across the session.
- enabling caching by default seems pretty reasonable, since the API & most datasets / etc., are only changing during the short period when they are 'under development'.
- one could implement more complicated logic (e.g., don't use caching for `:latest` version?) for individual calls (e.g., `get_file_by_id()` could detect that the user hasn't specified `use_cache=`, and adjust based on whatever is appropriate for that particular call).
- cached elements are stored on-disk for 30 days, so they are not permanent. 
- perhaps some effort might be invested in more explicit cache management, e.g., providing the user with a transparent way to flush the cache.

Please ensure the following before submitting a PR:

 - [x] if suggesting code changes or improvements, [open an issue](https://github.com/IQSS/dataverse-client-r/issues/new) first
 - [x] for all but trivial changes (e.g., typo fixes), add your name to [DESCRIPTION](https://github.com/IQSS/dataverse-client-r/blob/master/DESCRIPTION)
 - [x] for all but trivial changes (e.g., typo fixes), documentation your change in [NEWS.md](https://github.com/IQSS/dataverse-client-r/blob/master/NEWS.md) with a parenthetical reference to the issue number being addressed
 - [x] if changing documentation, edit files in `/R` not `/man` and run `devtools::document()` to update documentation
 - [x] add code or new test files to [`/tests`](https://github.com/IQSS/dataverse-client-r/tree/master/tests/testthat) for any new functionality or bug fix
 - [x] make sure `R CMD check` runs without error before submitting the PR

### CODE REVIEW

File: man-roxygen/dots.R [hunk 1]
```diff
@@ -1,3 +1,6 @@
-#' @param ... Additional arguments passed to an HTTP request function, such as
-#'   \code{\link[httr]{GET}}, \code{\link[httr]{POST}}, or
-#'   \code{\link[httr]{DELETE}}.
+#' @param ... Additional arguments passed to an HTTP request function,
+#'   such as \code{\link[httr]{GET}}, \code{\link[httr]{POST}}, or
+#'   \code{\link[httr]{DELETE}}. By default, HTTP requests use
+#'   values cached from previous identical calls. Use
+#'   \code{use_cache=FALSE} (or `Sys.setenv(DATAVERSE_USE_CACHE =
```
Comment (kuriwaki) [inline man-roxygen/dots.R:5]: Can we add `use_cache` as a documented argument in `get_file` as well? Currently it only shows up as a part of a the dot params arguments.  I want `use_cache` to be visible in `?get_dataframe_by_name`. That inherits the arguments of get_file via `@inheritDotParams get_file`  but that does not pull in the description of `cache`, so you don't see the cache explained in `?get_dataframe_by_name`

File: R/onload.R [hunk 2]
```diff
@@ -1,6 +1,27 @@
...
+    "api_cache"
+  )
+  get <- api_get_impl
+  if (!dir.exists(cache_directory)) {
+    status <- dir.create(cache_directory, recursive = TRUE)
+    if (!status)
+      warning("'dataverse' failed to create API cache")
+  }
+  if (dir.exists(cache_directory)) {
+    # disk cache with max age 30 days
```
Comment (kuriwaki) [inline R/onload.R:22]: Does this mean that after 30 days, the cache effectively disappears? We can document this somewhere.

File: R/onload.R [hunk 3]
```diff
@@ -1,6 +1,27 @@
+#' @importFrom cachem cache_disk
+#'
+#' @importFrom memoise memoise
 .onLoad <- function(libname, pkgname) {
-    # a <- Sys.getenv("DATAVERSE_SERVER")
-    # if(a == "") {
-    #     Sys.setenv("DATAVERSE_SERVER" = "dataverse.harvard.edu")
-    # }
+  # a <- Sys.getenv("DATAVERSE_SERVER")
+  # if(a == "") {
```
Comment (kuriwaki) [inline R/onload.R:6]: Probably delete this comment together, which seems to be a remnant of an old version. We now have an explicit error for missing `server` argument

File: R/onload.R (hunk 2 above)
Comment (mtmorgan) [reply R/onload.R:22]: items in the cache older than 30 days disappear, but items less than 30 days remain. I can update the documentation a bit; I wasn't sure what the strategy should be -- never, x days, to keep below y MB, etc.

**Comment by kuriwaki** [comment]
Re: 

> I imagine that a cache with an expiration date could be frustrating to users as it would make the runtime of their scripts unpredictable. Could we consider having it never clear, and providing a function that wipes the cache at the users request?

"never" sounds good to me. 30 days (or a year) is a decent default too, but I find myself returing to scripts that about a year old or more. Would be good to have a message in the beginning that explains this + how to cler the cache

**Comment by mtmorgan** [comment]
I'll update the pull request, but perhaps not until early next week

**Comment by kuriwaki** [comment]
Actually, after chatting with Ben I'd suggest we limit caching to when there is a dataset version is specified. (The default is not to specify). Say V1 is the latest version at time 1 and I cache that dataset with the default `version=latest`, and at time 2 there is a new version V2.  We wouldn't want the function to keep reading the cache silently.

**Comment by kuriwaki** [comment]
The `main` branch now has some new updates from #136 (including fixes to the actions check yaml) but it seems like I cannot synchronize someone else's fork with that easily. (I tried with rebase and push but I seem to have accidentally created a duplicate branch here instead of update this forked branch). So @mtmorgan, I'll let you synch with main if that's still doable.

**Comment by mtmorgan** [comment]
I think I can force push to the branch & PR; it would likely cause havoc for anyone with uncommitted changes. Shall I do that?

I can also push a commit that changes cache behavior

- `get_dataset_*()` have an explicit `use_cache = ` argument, default to TRUE except when `version = ":latest"` etc; disk cache never expires.
- cache other API calls on a per-session basis, addressing the speed issue in #112 
- basic cache management -- cache path, info (files & sizes), and reset (entire cache)

**Comment by mtmorgan** [comment]
I've force-pushed the branch, but am still updating the caching stuff.

**Comment by mtmorgan** [comment]
Sorry for the delay; I've added cache management functionality and implemented a (permanent) disk cache for explicitly versioned datasets, a session cache for other API calls, and the ability to turn off cache use entirely.

**Review by kuriwaki — APPROVED** [review]

**Comment by kuriwaki** [comment]
Thanks! This looks great and I've added  more documentation on the version argument, as well as mention this in the vignette. I've given this a new version number too. I'll plan on merging once tests pass.

**Pull request merged**